src/csharp/Grpc.Core.Api/Marshaller.cs

```diff
@@ -41,12 +47,91 @@ public Marshaller(Func<T, byte[]> serializer, Func<byte[], T> deserializer)
+                        // that we understand the intent of that method
+                        var parameters = bySegmentMethod.GetParameters();
+                        if (parameters.Length == 3 &&
+                            StringComparer.OrdinalIgnoreCase.Equals(parameters[1].Name, "offset")
+                            && (StringComparer.OrdinalIgnoreCase.Equals(parameters[2].Name, "length")
+                                || StringComparer.OrdinalIgnoreCase.Equals(parameters[2].Name, "count")))
+                        {
+                            var bySegmentTyped = (Func<byte[], int, int, T>)Delegate.CreateDelegate(typeof(Func<byte[], int, int, T>), deserializer.Target, bySegmentMethod);
+                            contextualDeserializer = ctx => {
+                                var len = ctx.PayloadLength;
+                                if (len == 0)
+                                {
+#if NET45
+                                    return bySegmentTyped(s_empty, 0, 0);
+#else
+                                    return bySegmentTyped(Array.Empty<byte>(), 0, 0);
+#endif
+                                }
+                                var ros = ctx.PayloadAsReadOnlySequence();
+                                if (ros.IsSingleSegment && MemoryMarshal.TryGetArray(ros.First, out var segment)
+                                    && segment.Count == len)
+                                {
+                                    // the data is *already* a single segment that is an array; use directly
+                                    return bySegmentTyped(segment.Array, segment.Offset, len);
+                                }
+                                else
+                                {
+                                    // lease a buffer from the pool, copy into that buffer, and deserialize
+                                    // from that, recycling the buffer when done
+                                    var arr = ArrayPool<byte>.Shared.Rent(len);
+                                    try
+                                    {
+                                        ros.CopyTo(arr);
+                                        return bySegmentTyped(arr, 0, len);
+                                    }
+                                    finally
+                                    {
+                                        ArrayPool<byte>.Shared.Return(arr);
+                                    }
+                                }
+                            };
+                            return true;
+                        }
+                    }
+                }
+            } catch { } // give up; best efforts only
+#endif
+            contextualDeserializer = default;
+            return false;
         }
 
+#if NET45
+        private static readonly byte[] s_empty = new byte[0]; // Array.Empty not available on all target platforms
```

Comment:
nit: Is it really worth adding `#if NET45` (which is not great for readability) as opposed to just using a static member always?

Reply:
fixed next push